lib/internal/modules/esm/hooks.js

```diff
@@ -497,13 +438,106 @@ class Hooks {
       __proto__: null,
       format,
       responseURL,
-      source,
+      source: ''+source,
     };
   }
 }
-
 ObjectSetPrototypeOf(Hooks.prototype, null);
 
+class HooksProxy {
+  /**
+   * The request & response segment of the shared memory. TextEncoder/Decoder (needed to convert
+   * requests & responses into a format supported by the comms channel) reads and writes with
+   * Uint8Array.
+   */
+  #data;
+  /**
+   * The lock/unlock segment of the shared memory. Atomics require this to be a Int32Array. This
+   * segment is used to tell the main to sleep when the worker is processing, and vice versa
+   * (for the worker to sleep whilst the main thread is processing).
+   * 0 -> main sleeps
+   * 1 -> worker sleeps
+   */
+  #lock;
+
+  #isReady = false;
+
+  /**
+   * @param {import('./loader.js).KeyedExports} customLoaders Exports from user-defined loaders
+   *  (as returned by `ModuleLoader.import()`).
+   */
+  constructor() {
+    const { InternalWorker } = require('internal/worker');
+
+    const data = new SharedArrayBuffer(204_800);
+    const lock = new SharedArrayBuffer(4);
+    this.#data = new Uint8Array(data);
+    this.#lock = new Int32Array(lock);
+
+    const worker = this.worker = new InternalWorker('internal/modules/esm/worker', {
+      stderr: false,
+      stdin: false,
+      stdout: false,
+      trackUnmanagedFds: false,
+      workerData: {
+        data,
+        lock,
+      },
+    });
+    worker.unref(); // ! Allows the process to eventually exit when worker is in its final sleep.
+  }
+
+  makeRequest(method, ...args) {
+    if (!this.#isReady) {
+      const { kIsOnline } = require('internal/worker');
+      if (!this.worker[kIsOnline]) {
+        Atomics.wait(this.#lock, 0, 0); // ! Block the main thread until the worker is ready.
+      }
+
+      this.#isReady = true;
+    }
+
+    TypedArrayPrototypeFill(this.#data, undefined); // Erase handled request/response data
+    const request = serialize({ method, args });
+    TypedArrayPrototypeSet(this.#data, request);
+
+    const chunks = [];
+    let done = false;
+    try {
+      while (done === false) {
+        this.#awaitResponse();
+        const chunkContainer = deserialize(this.#data);
+        // `deserialize` might return an error, such as a `RangeError`, rather than throw it
+        if (chunkContainer instanceof Error) { throw chunkContainer; }
+        ({ done } = chunkContainer);
+        const { value } = chunkContainer;
+        if (!value) { throw new ERR_INVALID_RETURN_VALUE('an object', method, undefined); }
+        ArrayPrototypePush(chunks, value);
```

Comment:
Did you try the following: pass the number of chunks in advance (in the `lock` SharedArrayBuffer?), and have a classic for loop iterating over them.

Reply:
I wish I’d seen all these comments before I just refactored the chunking logic 😄 Not really relevant now that I merged in the `postMessage` version.